python/mead/downloader.py

```diff
@@ -0,0 +1,300 @@
+    if not os.path.exists(cache_dir):
+        os.makedirs(cache_dir)
+    path_to_save_sha1 = os.path.join(cache_dir, sha1)
+    shutil.move(path_to_save, path_to_save_sha1)
+    print("downloaded data saved in {}".format(path_to_save_sha1))
+    return path_to_save_sha1
+
+
+def web_downloader(url):
+    import requests
+    import shutil
+    from baseline.progress import create_progress_bar
+    import os
+    r = requests.get(url, stream=True)
+    if r.status_code != 200:
+        raise RuntimeError("The file can not be downloaded")
+    path_to_save = "/tmp/data.dload-{}".format(os.getpid())
+    try:
+        print("downloading {}".format(url))
+        total_length = None
+        with open(path_to_save, 'wb') as f:
+            total_length = r.headers.get('content-length', None)
+            if total_length is not None:
+                total_length = int(total_length)
+                chunk_size = 1024
+                pg = create_progress_bar(total_length/chunk_size)
+                for chunk in r.iter_content(chunk_size=chunk_size):
+                    if chunk:
+                        f.write(chunk)
+                        pg.update()
+                        f.flush()
+                pg.done()
+            else:
+                raise RuntimeWarning("Total length can not be calculated, something wrong with the download file, trying ")
+                shutil.copyfileobj(r.raw, f)
+
+    except:  # this is too broad but there are too many exceptions to handle separately
+        raise RuntimeError("failed to download data from [url]: {} [to]: {}".format(url, path_to_save))
+    return path_to_save
+
+
+def validate_url(url):
+    import re
+    regex = re.compile(
+        r'^(?:http|ftp)s?://' # http:// or https://
+        r'(?:(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+(?:[A-Z]{2,6}\.?|[A-Z0-9-]{2,}\.?)|' #domain...
+        r'localhost|' #localhost...
+        r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})' # ...or ip
+        r'(?::\d+)?' # optional port
+        r'(?:/?|[/?]\S+)$', re.IGNORECASE)
+    return re.match(regex, url) is not None
+
+
+def update_cache(key):
+    import json
+    try:
+        dcaches = json.load(open(os.path.join(self.data_download_cache, "datasets-embeddings-cache.json")))[key]
+    except (IOError, KeyError):
+        return
+    del dcaches[key]
+    json.dump(dcaches, open(os.path.join(self.data_download_cache, "datasets-embeddings-cache.json"), "w"), indent=True)
+
+
+def delete_file(file_loc):
+    import os
+    import shutil
+    if os.path.exists(file_loc):
+        shutil.rmtree(file_loc)
+
+
+def check_sanity_file(file_loc, key=None):
+    import os
+    if os.path.exists(file_loc) and os.path.isfile(file_loc) and not mime_type(file_loc) == "text/html":  # dropbox doesn't give 404 in case the file does not exist, produces an HTML. Almost never the actual train/dev/test file is an HTML.
+        return True
+    else:
+        delete_file(file_loc)
```

Comment:
os.path.exists() is called twice when the file doesn't exist.

it's surprising to me that a method called `check_sanity_file` would delete something.